src/Simulation/Simulators/QuantumExecutor/EmptyQuantumExecutor.cs

```diff
@@ -0,0 +1,238 @@
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledR(IQArray<Qubit> controls, Pauli axis, double theta, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledR1(IQArray<Qubit> controls, double theta, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledR1Frac(IQArray<Qubit> controls, long numerator, long power, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledRFrac(IQArray<Qubit> controls, Pauli axis, long numerator, long power, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledS(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledSAdj(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledSWAP(IQArray<Qubit> controls, Qubit qubit1, Qubit qubit2)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledT(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledTAdj(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledX(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledY(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ControlledZ(IQArray<Qubit> controls, Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void Exp(IQArray<Pauli> paulis, double theta, IQArray<Qubit> qubits)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void ExpFrac(IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void H(Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Result M(Qubit qubit)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Result Measure(IQArray<Pauli> bases, IQArray<Qubit> qubits)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void OnAllocateQubits(IQArray<Qubit> qubits)
+        {
+        }
+
+        public virtual void OnBorrowQubits(IQArray<Qubit> qubits)
```

Comment:
Is there some way for the Executor to know how many Qubits were newly allocated vs. genuinely borrowed?

Reply:
Good suggestion for a new feature. I will address in another checkin.

---
In reply to: [349398743](https://github.com/microsoft/qsharp-runtime/pull/82#discussion_r349398743) [](ancestors = 349398743)